Debugger: List loaded IOP modules.

Ziemas opened the pull request:
![Screenshot_20250520_193501](https://github.com/user-attachments/assets/82250df3-56dc-4f28-8a39-90be8fcd0f1b)

(oops, text section end is broken in screenshot but fixed in pr)

### Description of Changes
Find and display the list of loaded modules by inspecting `loadcore`.

### Rationale behind Changes
It's nice for debugging to be able to quickly figure out what code is running.

### Suggested Testing Steps
Just open the list on whatever game/thing you can think of.

I'm quite worried about the loadcore detection and list traversal breaking if there is some possible struct layout that I didn't account for, or on something weird that doesn't have loadcore at all.

### Did you use AI to help find, test, or implement this issue or feature?
No

chaoticgd commented:
This looks good. The only problem I see is that `m_cpu.GetModuleList` is called every time `ModuleModel::data` is called. This information is not cached in the view, it's called every time it's redrawn, so I think it would make more sense to cache this data in the model and only update it every time there's a Refresh event (I'm thinking about merging the Refresh and VMUpdate events since I don't think they really serve separate purposes, it's just a hold-over from the old UI).

chaoticgd commented:
You could also see if you prefer it with [the section resize mode set to ResizeToContents](https://youtu.be/F8min4ZZJaI?si=xnhhlTO5TPVBGqCh&t=365).

chaoticgd commented:
It might be a good idea to add a hard upper limit to the iteration count in getIOPModules (getIOPThreads too) just in case the data structure gets corrupted somehow such as to create a cycle.

Ziemas commented:
> It might be a good idea to add a hard upper limit to the iteration count in getIOPModules (getIOPThreads too) just in case the data structure gets corrupted somehow such as to create a cycle.

Good idea, added some sanity checking to the list traversal.



> You could also see if you prefer it with [the vertical section resize mode set to ResizeToContents](https://youtu.be/F8min4ZZJaI?si=xnhhlTO5TPVBGqCh&t=365).

Nice, I added that to the module list and the thread list. Weirdly I don't see the effect of it on the EE thread list. (Does it only kick in after a certain amount of items?)



> This looks good. The only problem I see is that `m_cpu.GetModuleList` is called every time `ModuleModel::data` is called. This information is not cached in the view, it's called every time it's redrawn, so I think it would make more sense to cache this data in the model and only update it every time there's a Refresh event (I'm thinking about merging the Refresh and VMUpdate events since I don't think they really serve separate purposes, it's just a hold-over from the old UI).

I've now cached it in the model, also did that for the thread list which was what I copy pasted this all from. Also hooked up the refresh event to them. I really like the continuous updating :)

One thing though, now any selection the user makes in these windows get reset on every update, even when the system is paused. Idk if there's a way to solve that.

chaoticgd commented:
> Good idea, added some sanity checking to the list traversal.

100 seems a bit on the low side for the test, even if (almost all?) games will never load 100 at a time. Also remember to run clang-format.

> One thing though, now any selection the user makes in these windows get reset on every update, even when the system is paused. Idk if there's a way to solve that.

Oh right, I think that's why the VMUpdate and Refresh events were separate. You could either test if the old and new module lists are identical and not reset the model if that's the case, or just listen for the VMUpdate event only.

Ziemas commented:
I made it do nothing with the refresh event when paused, does that seem sensible? It seems to behave well in practice.
